pkg/controller/crdmirroring/crdmirroring_handlers.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2021 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crdmirroring
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+
+	"github.com/vmware-tanzu/antrea/pkg/controller/crdmirroring/types"
+)
+
+func (c *Controller) onNewCRDAdd(obj interface{}) {
+	curCRD := obj.(metav1.Object)
+
+	// Check if the legacy CRD mirroring current CRD exists.
+	if curCRD.GetAnnotations()[types.ManagedBy] == types.ControllerName {
+		klog.V(4).Infof("Processing mirroring %s %s/%s ADD event", c.crdName, curCRD.GetNamespace(), curCRD.GetName())
+		c.queueCRD(obj)
+	}
+}
+
+func (c *Controller) onNewCRDUpdate(prevObj, obj interface{}) {
+	curCRD := obj.(metav1.Object)
+
+	if curCRD.GetAnnotations()[types.ManagedBy] == types.ControllerName {
+		klog.Infof("Processing mirroring %s %s/%s UPDATE event", c.crdName, curCRD.GetNamespace(), curCRD.GetName())
+		c.queueCRD(obj)
+	}
+}
+
+func (c *Controller) onNewCRDDelete(obj interface{}) {
+	curCRD := getCRDFromDeleteAction(obj).(metav1.Object)
+
+	if curCRD.GetAnnotations()[types.ManagedBy] == types.ControllerName {
+		klog.Infof("Processing mirroring %s %s/%s DELETE event", c.crdName, curCRD.GetNamespace(), curCRD.GetName())
+		c.queueCRD(obj)
+	}
+}
+
+func (c *Controller) onLegacyCRDAdd(obj interface{}) {
+	legacyCRD := obj.(metav1.Object)
+
+	if legacyCRD.GetAnnotations()[types.StopMirror] != types.Stopped {
+		klog.Infof("Processing legacy %s %s/%s ADD event", c.crdName, legacyCRD.GetNamespace(), legacyCRD.GetName())
+		c.queueCRD(obj)
+	}
+}
+
+func (c *Controller) onLegacyCRDUpdate(prevObj, obj interface{}) {
+	legacyPrevCRD := prevObj.(metav1.Object)
+	legacyCRD := obj.(metav1.Object)
+
+	if legacyPrevCRD.GetAnnotations()[types.StopMirror] != types.Stopped {
+		klog.Infof("Processing legacy %s %s/%s UPDATE event", c.crdName, legacyCRD.GetNamespace(), legacyCRD.GetName())
+		c.queueCRD(obj)
+	}
+}
+
+func (c *Controller) onLegacyCRDDelete(obj interface{}) {
+	legacyCRD := getCRDFromDeleteAction(obj).(metav1.Object)
+
+	if legacyCRD.GetAnnotations()[types.StopMirror] != types.Stopped {
```

Comment:
this will panic if `getCRDFromDeleteAction` returns nil.
I meant the error was processed twice in the caller and callee, but it still needs to check nil if the callee may return nil.

Reply:
I got what you meant now.